sdks/python/apache_beam/runners/portability/universal_local_runner.py

```diff
@@ -0,0 +1,398 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import functools
+import logging
+import os
+import Queue as queue
+import socket
+import subprocess
+import sys
+import threading
+import time
+import traceback
+import uuid
+from concurrent import futures
+
+import grpc
+from google.protobuf import text_format
+
+from apache_beam.portability.api import beam_fn_api_pb2
+from apache_beam.portability.api import beam_job_api_pb2
+from apache_beam.portability.api import beam_job_api_pb2_grpc
+from apache_beam.runners import runner
+from apache_beam.runners.portability import fn_api_runner
+
+TERMINAL_STATES = [
+    beam_job_api_pb2.JobState.DONE,
+    beam_job_api_pb2.JobState.STOPPED,
+    beam_job_api_pb2.JobState.FAILED,
+    beam_job_api_pb2.JobState.CANCELLED,
+]
+
+
+class UniversalLocalRunner(runner.PipelineRunner):
+  """A BeamRunner that executes pipelines via the Beam Job API.
```

Comment:
<!--new_thread; commit:77e51af3ee2af39264c366a2602957e7e9bb20c7; resolved:0-->
Perhaps we should say '...executes Python pipelines...' given that we hardcode python SDK harness? Do we expect to support other SDK harnesses it in the future in this runner? If so, maybe we should clarify it in a TODO?

Reply:
Done. Given that this takes a Python pipeline object I don't think we'd use this Python API for other langauges. (We'd use the job API or the language native API if we wanted to do that.)